# HG changeset patch
# User Mike Miller <mtmiller@ieee.org>
# Date 1376451750 14400
#      Tue Aug 13 23:42:30 2013 -0400
# Node ID 3f4091dddc13088cb535be42a2f8c9b7dbef117c
# Parent  03a666018e0f2bed9d53da61166aa82c56dcfd21
doc: Update doc cache delimiter to work with Texinfo 5.0 (bug #39778)

* doc/interpreter/mk_doc_cache.m: Change doc delimiter to ASCII 30
(record separator) because Texinfo 5.0 uses ASCII 31 (unit separator).
Use hexadecimal representation for consistency with other scripts.
* libinterp/corefcn/help.cc (install_built_in_docstrings): Change doc
delimiter to ASCII 30 (record separator).
* doc/interpreter/munge-texi.pl, libinterp/gendoc.pl, scripts/mkdoc.pl:
Change doc delimiter to ASCII 30 (record separator). Use hexadecimal escape
sequence for consistency and to avoid using literal control characters.

diff --git a/doc/interpreter/mk_doc_cache.m b/doc/interpreter/mk_doc_cache.m
--- a/doc/interpreter/mk_doc_cache.m
+++ b/doc/interpreter/mk_doc_cache.m
@@ -19,17 +19,17 @@ args = argv ();
 if (nargin < 2)
   error ("usage: mk_doc_cache OUTPUT-FILE DOCSTRINGS-FILE ...");
 endif
 
 output_file = args{1};
 docstrings_files = args(2:end);
 
 ## Special character used as break between DOCSTRINGS
-doc_delim = char (31);
+doc_delim = char (0x1e);
 
 ## Read the contents of all the DOCSTRINGS files into TEXT.
 ## It is more efficient to fork to shell for makeinfo only once on large data
 
 nfiles = numel (docstrings_files);
 text = cell (1, nfiles);
 for i = 1:nfiles
   file = docstrings_files{i};
@@ -47,17 +47,17 @@ for i = 1:nfiles
       [~, text{i}] = strtok (tmp, doc_delim);
     endif
   endif
 endfor
 text = [text{:}, doc_delim];
 
 ## Strip Texinfo markers and docstring separators.
 text = regexprep (text, "-\\*- texinfo -\\*-[ \t]*[\r\n]*", "");
-text = strrep (text, '@', "@@");
+text = strrep (text, [doc_delim "@"], [doc_delim "@@"]);
 
 ## Write data to temporary file for input to makeinfo
 [fid, name, msg] = mkstemp ("octave_doc_XXXXXX", true);
 if (fid < 0)
   error ("%s: %s\n", name, msg);
 endif
 fwrite (fid, text, "char");
 fclose (fid);
diff --git a/doc/interpreter/munge-texi.pl b/doc/interpreter/munge-texi.pl
--- a/doc/interpreter/munge-texi.pl
+++ b/doc/interpreter/munge-texi.pl
@@ -1,17 +1,17 @@
 #!/usr/bin/perl -w
 
 # Validate program call
 die "usage: munge-texi TOP-SRCDIR DOCSTRING-FILE < file" if (@ARGV < 2);
 
 $top_srcdir = shift (@ARGV);
 
 # Constant patterns
-$doc_delim = qr/^\c_/;
+$doc_delim = qr/^\x{1e}/;
 $tex_delim = qr/\Q-*- texinfo -*-\E/;
 $comment_line = qr/^\s*(?:$|#)/;
 # Pre-declare hash size for efficiency
 keys(%help_text) = 1800;
 
 ################################################################################
 # Load DOCSTRINGS into memory while expanding @seealso references
 foreach $DOCSTRING_file (@ARGV)
diff --git a/libinterp/corefcn/help.cc b/libinterp/corefcn/help.cc
--- a/libinterp/corefcn/help.cc
+++ b/libinterp/corefcn/help.cc
@@ -962,17 +962,17 @@ install_built_in_docstrings (void)
 {
   std::string fname = Vbuilt_in_docstrings_file;
 
   std::ifstream file (fname.c_str (), std::ios::in | std::ios::binary);
 
   if (file)
     {
       // Ignore header;
-      file.ignore (1000, 0x1f);
+      file.ignore (1000, 0x1e);
 
       if (file.gcount () == 1000)
         {
           // We use std::cerr here instead of calling Octave's warning
           // function because install_built_in_docstrings is called
           // before the interpreter is initialized, so warning messages
           // won't work properly.
 
@@ -983,17 +983,17 @@ install_built_in_docstrings (void)
 
       // FIXME -- eliminate fixed buffer size.
       size_t bufsize = 100000;
 
       OCTAVE_LOCAL_BUFFER (char, buf, bufsize);
 
       while (! file.eof ())
         {
-          file.getline (buf, bufsize, 0x1f);
+          file.getline (buf, bufsize, 0x1e);
 
           std::string tmp (buf);
 
           size_t pos = tmp.find ('\n');
 
           std::string fcn = tmp.substr (0, pos);
 
           octave_value ov = symbol_table::find_built_in_function (fcn);
diff --git a/libinterp/gendoc.pl b/libinterp/gendoc.pl
--- a/libinterp/gendoc.pl
+++ b/libinterp/gendoc.pl
@@ -59,15 +59,15 @@ DFFILE: foreach $df_fname (@ARGV)
     }
   }
   close (DF_FH);
 
   ## Print results in DOCSTRING format
   foreach $i (0 .. $#func_list)
   {
     $func = $func_list[$i];
-    print "$func\n";
+    print "\x{1e}$func\n";
     print "\@c $func $src_fname\n";
     print $docstr[$i],"\n";
   }
 
 }
 
diff --git a/scripts/mkdoc.pl b/scripts/mkdoc.pl
--- a/scripts/mkdoc.pl
+++ b/scripts/mkdoc.pl
@@ -47,17 +47,17 @@ MFILE: foreach $m_fname (@ARGV)
   if ($2)
     { $fcn = "$2$3/$4"; }
   else
     { $fcn = $4; }
 
   @help_txt = gethelp ($fcn, $full_fname);
   next MFILE if ($help_txt[0] eq "");
 
-  print "$fcn\n";
+  print "\x{1e}$fcn\n";
   print "\@c $fcn scripts/$m_fname\n";
 
   foreach $_ (@help_txt)
   {
     s/^\s+\@/\@/ unless $in_example;
     s/^\s+\@group/\@group/;
     s/^\s+\@end\s+group/\@end group/;
     $in_example = (/\s*\@example\b/ .. /\s*\@end\s+example\b/);
